File: PrivateVault.ai-main/sdk/python/privatevault_sdk/verify.py
import hashlib
import os
from typing import List, Tuple


def verify_manifest(bundle_path: str) -> Tuple[bool, List[str]]:
    manifest_path = os.path.join(bundle_path, "hashes", "sha256sum.txt")
    if not os.path.exists(manifest_path):
        return False, ["MANIFEST_MISSING"]

    mismatches = []
    with open(manifest_path, "r", encoding="utf-8") as f:
        lines = [line.strip() for line in f if line.strip()]

    for line in lines:
        expected, relpath = line.split("  ", 1)
        file_path = os.path.join(bundle_path, relpath)
        if not os.path.exists(file_path):
            mismatches.append(relpath)
            continue
        digest = hashlib.sha256()
        with open(file_path, "rb") as fh:
            for chunk in iter(lambda: fh.read(65536), b""):
                digest.update(chunk)
        if digest.hexdigest() != expected:
            mismatches.append(relpath)

    return len(mismatches) == 0, mismatches

File: PrivateVault.ai-main/sdk/node/src/verify.js
import crypto from 'crypto';
import fs from 'fs';
import path from 'path';

export function verifyManifest(bundlePath) {
  const manifestPath = path.join(bundlePath, 'hashes', 'sha256sum.txt');
  if (!fs.existsSync(manifestPath)) {
    return { valid: false, mismatches: ['MANIFEST_MISSING'] };
  }
  const lines = fs.readFileSync(manifestPath, 'utf-8').split('\n').filter(Boolean);
  const mismatches = [];
  for (const line of lines) {
    const [expected, relpath] = line.split('  ');
    const filePath = path.join(bundlePath, relpath);
    if (!fs.existsSync(filePath)) {
      mismatches.push(relpath);
      continue;
    }
    const data = fs.readFileSync(filePath);
    const digest = crypto.createHash('sha256').update(data).digest('hex');
    if (digest !== expected) {
      mismatches.push(relpath);
    }
  }
  return { valid: mismatches.length === 0, mismatches };
}
